model/riscv_sys_control.sail

```diff
@@ -162,22 +174,24 @@ function check_CSR_access(csrrw, csrpr, p, isWrite) =
 function check_TVM_SATP(csr : csreg, p : Privilege) -> bool =
   not(csr == 0x180 & p == Supervisor & mstatus[TVM] == 0b1)
 
-function check_Counteren(csr : csreg, p : Privilege) -> bool =
-  match(csr, p) {
-    (0xC00, Supervisor) => mcounteren[CY] == 0b1,
-    (0xC01, Supervisor) => mcounteren[TM] == 0b1,
-    (0xC02, Supervisor) => mcounteren[IR] == 0b1,
+// There are several features that are controlled by machine/supervisor enable
+// bits (m/senvcfg, m/scounteren, etc.). This abstracts that logic.
+function mode_bit_enabled(p : Privilege, machine_bit : bit, supervisor_bit : bit) -> bool = match p {
+  Machine => true,
+  Supervisor => machine_bit == bitone,
+  User => machine_bit == bitone & (not(haveSupMode()) | supervisor_bit == bitone),
+}
 
-    (0xC00, User) => mcounteren[CY] == 0b1 & (not(haveSupMode()) | scounteren[CY] == 0b1),
-    (0xC01, User) => mcounteren[TM] == 0b1 & (not(haveSupMode()) | scounteren[TM] == 0b1),
-    (0xC02, User) => mcounteren[IR] == 0b1 & (not(haveSupMode()) | scounteren[IR] == 0b1),
-
-    (_, _) => /* no HPM counters for now */
-              if   0xC03 <=_u csr & csr <=_u 0xC1F
-              then false
-              else true
-  }
+// Return true if the counter is enabled OR the CSR is not a counter.
+function check_Counteren(csr : csreg, p : Privilege) -> bool = {
+  // Check if it is not a counter.
+  if csr <_u 0xC00 | csr >_u 0xC1F
```

Comment:
Maybe this should be a fatal error instead? Do we even call this for non-counter CSRs?

Reply:
Yeah it's called for all CSRs - see `check_CSR()` in `riscv_sys_control.sail`.

Reply:
Since the same style is used for other checks I guess we keep it as is. But IMO it would be cleaner to push that check into check_CSR, so we don't have all these checks being called for every CSR.